File: src/main/java/com/core/base/corebase/support/JwtInterceptor.java
package com.core.base.corebase.support;

import com.core.base.corebase.common.code.ErrorCode;
import com.core.base.corebase.common.exception.BaseException;
import com.core.base.corebase.config.AuthenticationFacade;
import com.core.base.corebase.repository.UserRepository;
import io.jsonwebtoken.Claims;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.http.HttpMethod;
import org.springframework.web.servlet.HandlerInterceptor;

public class JwtInterceptor implements HandlerInterceptor {

    private final AuthenticationFacade authenticationFacade;
    private final JwtProvider jwtProvider;
    private final UserRepository userRepository;

    public JwtInterceptor(AuthenticationFacade authenticationFacade,
                         JwtProvider jwtProvider,
                         UserRepository userRepository) {
        this.authenticationFacade = authenticationFacade;
        this.jwtProvider = jwtProvider;
        this.userRepository = userRepository;
    }

    @Override
    public boolean preHandle(HttpServletRequest request,
                            HttpServletResponse response,
                            Object handler) throws Exception {

        if (!HttpMethod.OPTIONS.matches(request.getMethod())) {
            String authHeader = request.getHeader("Authorization");
            if (authHeader != null && authHeader.startsWith("Bearer ")) {
                String token = authHeader.substring(7);
                Claims claims = jwtProvider.getBody(token);

                if (jwtProvider.isAccess(claims)) {
                    userRepository.findByUid(jwtProvider.getId(claims))
                        .ifPresent(user -> authenticationFacade.setInfo(
                            user.getUid(),
                            user.getEmail(),
                            user.getName()
                        ));
                    return true;
                }
            }
            throw new BaseException(ErrorCode.INVALID_TOKEN);
        }
        return true;
    }
}

File: src/main/kotlin/com/core/base/corebase/support/JwtInterceptor.kt
package com.core.base.corebase.support

import com.core.base.corebase.common.code.ErrorCode
import com.core.base.corebase.common.exception.BaseException
import com.core.base.corebase.config.AuthenticationFacade
import com.core.base.corebase.repository.AccountRepository
import com.core.base.corebase.repository.UserRepository
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.data.repository.findByIdOrNull
import org.springframework.http.HttpMethod
import org.springframework.web.servlet.HandlerInterceptor

class JwtInterceptor(
    private val authenticationFacade: AuthenticationFacade,
    private val jwtProvider: JwtProvider,
    private val userRepository: UserRepository
) : HandlerInterceptor {

    override fun preHandle(request: HttpServletRequest, response: HttpServletResponse, handler: Any): Boolean {
        if (!HttpMethod.OPTIONS.matches(request.method))
            request.getHeader("Authorization")
                ?.takeIf { it.startsWith("Bearer ") }
                ?.substring(7)
                ?.let { jwtProvider.getBody(it) }
                ?.takeIf { jwtProvider.isAccess(it) }
                ?.let { jwtProvider.getId(it) }
                ?.let { userRepository.findByUid(it) }
                ?.apply { authenticationFacade.setInfo(uid, email, name) }
                ?: throw BaseException(ErrorCode.INVALID_TOKEN)
        return super.preHandle(request, response, handler)
    }

}